File: ObjC/8_UIResponder/SampleApp/ViewController.m
#import "ViewController.h"

// UIReponsder Chain(Chain of Responsibility)
//  - Touch 이벤트
//   UIView -> UIViewController -> UIWindow -> AppDelegate

//  => 반드시 Chain을 따르는 모든 것이 UIReponder의 자식이어야 한다.

@implementation ViewController

- (IBAction)buttonClick:(id)sender {
  NSLog(@"buttonClick");
}

- (IBAction)tap:(id)sender {
   NSLog(@"tap");
}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view from its nib.
}

#if 0
- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
  NSLog(@"touchesBegan");
  
  NSSet* t = event.allTouches;
  NSLog(@"%lu개의 터치가 발생", t.count);
  [super touchesBegan:touches withEvent:event];
}

- (void)touchesEnded:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
  NSLog(@"touchesEnded");
  [super touchesEnded:touches withEvent:event];
}

- (void)touchesMoved:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
  NSLog(@"touchesMoved");
  [super touchesMoved:touches withEvent:event];
}

- (void)touchesCancelled:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
  NSLog(@"touchesCancelled");
}
#endif


@end

File: Swift/7_NotificationCenter/SampleApp/ViewController.swift
import UIKit

// Soft Keyboard에 의해서 TextField가 가려질 경우
// 처리하는 방법
//  => NotificationCenter
class ViewController: UIViewController {
  @IBOutlet var emailField: UITextField!
  @IBOutlet var passwordField: UITextField!
  
  @IBOutlet var bottomConstraint: NSLayoutConstraint!
  
  // 비동기(Async)
  //  => 결과를 처리하기 위해서 '콜백'을 등록한다.
  //  => 비동기 루틴이 어떤 컨텍스트에서 수행되는지 설정할 수 있다.
  //   - UI에 대한 처리 반드시 main 스레드에서 수행이 되어야 합니다.
  //   - 그 외 작업은 다른 스레드에서 처리하는 것이 가능합니다.
  //    : DispatchQueue
  
  override func viewDidLoad() {
    super.viewDidLoad()
    // Do any additional setup after loading the view.
    
    let notificationCenter = NotificationCenter.default
    
    // 사용자 정의 알림을 등록하는 것도 가능합니다.
    notificationCenter.addObserver(forName: NSNotification.Name(rawValue: "RECEIVESMS"), object: nil, queue: .main) { _ in
      print("메일이 도착하였습니다.")
    }
    
    notificationCenter.addObserver(forName: UIResponder.keyboardWillShowNotification,
                                   object: nil,
                                   queue: .main) { notification in
                                    
      let keyboardHeight = self.softKeyboardHeight(notification)
      
      UIView.animate(withDuration: 1, animations: {
        self.bottomConstraint.constant = 32 + keyboardHeight
        self.view.layoutIfNeeded()
      })
    }
    
    notificationCenter.addObserver(forName: UIResponder.keyboardWillHideNotification,
                                   object: nil,
                                   queue: .main) { _ in
      print("keyboardWillHideNotification")
      UIView.animate(withDuration: 1, animations: {
        self.bottomConstraint.constant = 32
        self.view.layoutIfNeeded()
      })
    }
    
    // 배터리 상태를 모니터링 하기 위해서는
    let device = UIDevice.current
    device.isBatteryMonitoringEnabled = true // !!
    
    notificationCenter.addObserver(forName: UIDevice.batteryLevelDidChangeNotification, object: nil, queue: .main) { notification in
      print("batteryLevelDidChangeNotification: \(notification)")
    }
    notificationCenter.addObserver(forName: UIDevice.batteryStateDidChangeNotification, object: nil, queue: .main) { notification in
      print("batteryStateDidChangeNotification: \(notification)")
    }
    
    /*
     notificationCenter.addObserver(self,
                                    selector: #selector(keyboardWillShow),
                                    name: UIResponder.keyboardWillShowNotification,
                                    object: nil)
     notificationCenter.addObserver(self,
                                    selector: #selector(keyboardWillHide),
                                    name: UIResponder.keyboardWillHideNotification,
                                    object: nil)
     */
  }
  
  @objc func keyboardWillShow(_ notification: Notification) {
    print("keyboardWillShow")
    let keyboardHeight = softKeyboardHeight(notification)
    bottomConstraint.constant = 32 + keyboardHeight
  }
  
  @objc func keyboardWillHide(_ notification: Notification) {
    print("keyboardWillHide")
    bottomConstraint.constant = 32
  }
  
  // 키보드의 frame 정보를 알아내기 위해서는, notification의 정보가 필요합니다.
  func softKeyboardHeight(_ notification: Notification) -> CGFloat {
    if let frame = notification.userInfo?[UIResponder.keyboardFrameEndUserInfoKey] as? NSValue {
      return frame.cgRectValue.height
    }
    
    return 0
  }
  
  @IBAction func click(_ sender: Any) {
    let notificationCenter = NotificationCenter.default
    notificationCenter.post(name: NSNotification.Name(rawValue: "RECEIVESMS"), object: nil)
  }
  
  override func touchesEnded(_ touches: Set<UITouch>,
                             with event: UIEvent?)
  {
    view.endEditing(true)
  }
}
